File: 377-combination_sum_iv.py
# opt 1) brute force: vector length * cartesian product, filter
# optimization: sum with abort upon impossible scenario
# -> TLE

# opt 2) backtrack

# time: O(t*n)
# space: O(t)
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        # potential optimization
        # nums.sort()

        @functools.lru_cache(maxsize = None)
        def combs(remain):
            if remain == 0:
                return 1

            result = 0
            for num in nums:
                if remain - num >= 0:
                    result += combs(remain - num)
                # potential optimization
                # else:
                #     break

            return result

        return combs(target)

File: 377-combination_sum_iv.cs
public class Solution {
    public int CombinationSum4(int[] nums, int target)
    {
        var memo = new Dictionary<int, int>();
        return dfs(target);

        int dfs(int remain)
        {
            if (memo.ContainsKey(remain)) {
                return memo[remain];
            }

            if (remain == 0) {
                return 1;
            }
            if (remain < 0)
            {
                return 0;
            }

            int cnt = 0;
            foreach(int num in nums)
            {
                cnt += dfs(remain - num);
            }

            memo[remain] = cnt;

            return cnt;
        }
    }
}
